networkx/algorithms/tsp.py

```diff
@@ -0,0 +1,319 @@
+"""
+================================
+Travelling Salesman Problem (TSP)
+================================
+
+Implementation of heuristic and metaheuristic algorithms
+for solving and approximating the TSP problem.
+
+Categories of algorithms which are implemented:
+- Greedy
+- Simulated Annealing (SA)
+- Threshold Accepting (TA)
+- Tabu Search
+- Ant Colony Optimization (ACO)
+
+Travelling Salesman Problem tries to find, given the weight
+(distance) between all points where salesman has to visit, the
+route so that:
+- Total distance (cost) which salesman travels to be minimized.
+- Salesman has to return to the point where he stated.
+- Salesman has to visit each point only once apart fom source
+point.
+
+It is an NP-hard problem in combinatorial optimization,
+important in operations research and theoretical computer science.
+
+http://en.wikipedia.org/wiki/Travelling_salesman_problem
+"""
+import math
+from operator import itemgetter
+from random import choice, randint, random
+import networkx as nx
+
+__all__ = ['greedy_tsp', 'simulated_annealing_tsp']
+
+
+def greedy_tsp(G, source, weight='weight'):
+    """Finds the route that salesman has to visit in order
+    to minimize total distance and total distance using a
+    simple greedy algorithm.
+
+    In essence, function returns a cycle given a source point
+    which total cost of cycle is minimized.
+
+    Parameters
+    ----------
+    G: A complete weighted graph.
+
+    source : node
+        Starting node
+
+    weight: string, optional (default='weight')
+        Edge data key corresponding to the edge weight
+
+    Raises
+    ------
+    NetworkXError
+        If graph is not either complete or weighed,
+        algorithm raises an exception.
+
+    Returns
+    -------
+    sol : list, cost : float
+        Returns the route (list of nodes) that salesman
+        has to follow to minimize total cost, and total
+        cost of algorithm's solution.
+
+    Examples
+    --------
+    >>> import networkx as nx
+    >>> G = nx.DiGraph()
+    >>> G.add_weighted_edges_from({('A', 'B', 3),
+    ...                           ('A', 'C', 17), ('A', 'D', 14), ('B', 'A', 3),
+    ...                          ('B', 'C', 12), ('B', 'D', 16), ('C', 'A', 13),
+    ...                           ('C', 'B', 12), ('C', 'D', 4), ('D', 'A', 14),
+    ...                           ('D', 'B', 15), ('D', 'C', 2)})
+    >>> sol = nx.greedy_tsp(G, 'D')
```

Comment:
Use `cycle, weight = nx.greedy_tsp(G, 'D')` instead to make the return values clearer.